### PR TITLE
Elastic Search API Client

### PR DESCRIPTION
**Pull request by indragiek**
Implements #25.

`ElasticSearchAPIClient` is the class that handles all communication between ElasticSearch and our client application. Here's a quick run down of what's in here:

`ElasticSearchAPIClient` contains simple `get()`, `add()`, `update()`, and `delete()` methods, corresponding to the `GET`, `POST`, `PUT`, and `DELETE` HTTP methods. All of these methods return an instance of `APICall<T>`.

`APICall<T>` encapsulates a single HTTP request. Encapsulating this in a separate class is beneficial because it allows for delayed execution of these requests and also retry, in case the request fails. For example, to implement the offline functionality, we can queue up these objects for executing later if it is determined that the phone does not have internet connectivity.

`APICall<T>` provides two ways to execute the request. One method is the `execute()` method, which executes the request synchronously on the current thread. This is **only used** for unit tests, because testing asynchronous code in unit tests is troublesome (at least with our current testing setup). This method will synchronously execute the request and return the result (an ElasticSearch document) or raise one of several exceptions if the request could not be completed for any reason. This should not be called from application code because we never want to block the main thread while waiting for a response to the request, since the network is unpredictable and may take hundreds of ms to return a response.

The method on `APICall<T>` that we will be using in the app is `enqueue()`, which is asynchronous and allows you to provide an object that implements `APICallback<T>` that will be notified when the request succeeds or fails.

### CODE REVIEW

File: Team9Project/src/com/indragie/cmput301as1/ElasticSearchAPIClient.java [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+		
+		//================================================================================
+		// API
+		//================================================================================
+		/**
+		 * Cancels the request, if possible.
+		 */
+		public void cancel() {
+			call.cancel();
+		}
```
Comment (mrthegreensky) [inline Team9Project/src/com/indragie/cmput301as1/ElasticSearchAPIClient.java:192]: Is cancel used somewhere?

Comment (indragiek) [inline Team9Project/src/com/indragie/cmput301as1/ElasticSearchAPIClient.java:192]: Not at the moment, its more of a hypothetical use case if we did want to cancel an in-progress request. I'm not sure if that'll come up anywhere in this app though.

Comment (mrthegreensky) [inline Team9Project/src/com/indragie/cmput301as1/ElasticSearchAPIClient.java:192]: Okay, just making sure. It just came by as a bit weird

**Comment by indragiek** [comment]
Is this OK? Does anyone else have comments? @CMPUT301W15T09/owners

**Comment by williums** [comment]
Everything looks fine to me

**Pull request merged**